File: client/src/sim/mapgen.ts
/** 결정론적 초기 맵 생성 — docs/mapgen.md. */

import { clampInt, floorDiv, hash32, iabs } from "./intmath.ts";
import {
  BEDROCK,
  EMPTY,
  H,
  N,
  ROCK,
  SAND,
  SCREE,
  SOIL,
  W,
} from "./terrain.ts";

export const MAPGEN_VERSION = 3;
export const NOISE_SHIFT = 7;
export const SURFACE_BASE = 262;
export const SURFACE_AMP = 70;
export const BEDROCK_Y = 522;
export const SPAWN_MIN_GAP = 36;
export const SPAWN_MAX_RELIEF = 12;

const NOISE_MASK = 127;
const NOISE_SALT = 0x4d47;
const ANCHOR_SALT = 0xa11c;
const ARCH_SALT = 0xa2c4;
const ANCHOR_BASE_X = [160, 320, 640, 800] as const;

function setCell(target: Uint8Array, x: number, y: number, material: number): void {
  target[y * W + x] = material;
}

function fillVertical(
  target: Uint8Array,
  x: number,
  yStart: number,
  yEnd: number,
  material: number,
): void {
  const start = clampInt(yStart, 0, H - 1);
  const end = clampInt(yEnd, 0, H - 1);
  for (let y = start; y <= end; y++) setCell(target, x, y, material);
}

function peakAt(column: number, center: number, width: number, amplitude: number): number {
  return floorDiv(clampInt(width - iabs(column - center), 0, width) * amplitude, width);
}

function generatedSurface(mapSeed: number, column: number): number {
  const sample = column >> NOISE_SHIFT;
  const fraction = column & NOISE_MASK;
  const start = hash32(mapSeed, sample, NOISE_SALT, 0) & 0xffff;
  const end = hash32(mapSeed, sample + 1, NOISE_SALT, 0) & 0xffff;
  const noise = start + (((end - start) * fraction) >> NOISE_SHIFT);
  const mainSeed = hash32(mapSeed, 0, 0x504b, 0);
  const sideSeed = hash32(mapSeed, 1, 0x504b, 0);
  const valleySeed = hash32(mapSeed, 2, 0x504b, 0);
  const mainCenter = clampInt(240 + (mainSeed & 511), 240, 720);
  const sideCenter = (mainCenter < 480 ? 770 : 190) + ((sideSeed >>> 16) & 63) - 31;
  const mainPeak = peakAt(column, mainCenter, 145 + ((mainSeed >>> 9) & 127), 148 + ((mainSeed >>> 16) & 63));
  const sidePeak = peakAt(column, sideCenter, 120 + (sideSeed & 63), 76 + ((sideSeed >>> 6) & 47));
  const valley = peakAt(column, floorDiv(mainCenter + sideCenter, 2), 110 + (valleySeed & 63), 32 + ((valleySeed >>> 6) & 31));
  return clampInt(SURFACE_BASE + (((noise - 32768) * SURFACE_AMP) >> 16) - mainPeak - sidePeak + valley, 72, 342);
}

/* ══ 지질 구역 (§3.1) ═════════════════════════════════════════════════════
   맵을 좌우로 나눠 **구역마다 다른 지층**을 깐다.

   그 전에는 960열 전부가 같은 지층(SAND 18 / SOIL 28 / SAND 10 / SCREE 12 / SOIL 40)
   이었다. 그래서 실측으로 갈라놓은 안식각 26.6°/40.3°/44.2° 가 전술적으로 아무 일도
   하지 않았다 — 어디를 파도 같은 순서로 같은 것이 나오니 **서 있는 자리가 의미가 없었다.**

   모래 분지를 파면 26.6° 로 넓게 흘러내리고, 점토 대지를 파면 40.3° 벽이 그대로 선다. */

/** 구역 프로파일. 위에서부터의 두께이고, 남는 깊이는 `ROCK` 이 채운다. */
interface Province {
  /** [SAND, SOIL, SAND, SCREE, SOIL] 두께 (셀) */
  bands: readonly [number, number, number, number, number];
  /** 기반암 상단을 지표에서 얼마나 아래에 둘 것인가. 0 이면 `BEDROCK_Y` 그대로 */
  bedrockDepth: number;
}

/* **표면 재질이 구역마다 달라야 한다.** 첫 판에서는 네 구역 모두 SAND 로 시작했는데,
   그러면 화면상 전부 모래라 플레이어가 서 있는 땅을 구분할 수 없다 — 지질이 달라도
   보이지 않으면 전술이 되지 않는다. 앞 밴드를 0 으로 두어 아래 재질을 노출시킨다.
   노출된 SOIL(40.3°)·SCREE(44.2°) 는 지표 경사(최대 25°)보다 안식각이 급해서
   그대로 서 있는다 — 노출시켜도 정착 비용이 안 든다 (실측 §11). */
export const PROVINCES: readonly Province[] = [
  /* 0 모래 분지 — 두꺼운 모래가 지표에 드러난다. 파면 26.6° 로 넓게 흘러내린다 */
  { bands: [34, 14, 6, 10, 30], bedrockDepth: 0 },
  /* 1 점토 대지 — 점토가 지표다. 40.3° 급한 벽이 서고 잘 안 무너진다 */
  { bands: [0, 48, 4, 8, 40], bedrockDepth: 0 },
  /* 2 자갈 사면 — 자갈이 지표다. 44.2° 각진 더미를 만든다 */
  { bands: [0, 0, 0, 34, 30], bedrockDepth: 0 },
  /* 3 암반 선반 — 얇은 표토 밑이 바로 기반암이다. 깊이 못 파는 대신 발밑이 안전하다 */
  { bands: [6, 12, 0, 6, 16], bedrockDepth: 132 },
];

const PROVINCE_SALT = 0x9e07;
/** 구역 배치 순열. **한 맵 안에 네 지질이 전부 나오는 것을 보장한다.**

    처음에는 구역마다 `hash32(...) % 4` 로 독립 추첨했는데, 그러면 시드에 따라 맵 전체가
    한 지질로 덮인다 (실측: 모래 809열 / 960). 그러면 "어디에 서 있느냐"가 다시 의미를
    잃는다 — 다양성이 이 변경의 전부인데 우연에 맡길 이유가 없다.

    정수 전용 Fisher-Yates 다. 결정론을 위해 `hash32` 만 쓴다. */
function provinceOrder(mapSeed: number): number[] {
  const order = [0, 1, 2, 3];
  for (let i = order.length - 1; i > 0; i--) {
    const j = hash32(mapSeed, i, PROVINCE_SALT, 1) % (i + 1);
    const t = order[i];
    order[i] = order[j];
    order[j] = t;
  }
  return order;
}

/** 구역 경계가 지층을 수직으로 자르지 않게 섞는 폭 (셀) */
export const PROVINCE_BLEND = 48;

/** 열 `x` 가 속한 구역과, 경계에서의 이웃·혼합 가중치 */
function provinceAt(mapSeed: number, x: number): { a: number; b: number; t: number } {
  const count = 4 + (hash32(mapSeed, 0, PROVINCE_SALT, 0) & 1); // 4 또는 5 구역
  const width = floorDiv(W, count);
  let index = floorDiv(x, width);
  if (index >= count) index = count - 1;
  const localX = x - index * width;

  const pick = (i: number): number => provinceOrder(mapSeed)[i % PROVINCES.length];

  const here = pick(index);
  if (localX >= PROVINCE_BLEND || index === 0) return { a: here, b: here, t: 0 };
  /* 경계 왼쪽 절반은 이전 구역과 섞는다. t 는 0..PROVINCE_BLEND 의 정수 가중치 */
  return { a: pick(index - 1), b: here, t: localX };
}

/** 두 프로파일을 정수 가중 평균한다. `t/PROVINCE_BLEND` 비율로 b 쪽에 간다 */
function blendBand(a: number, b: number, t: number): number {
  return a + floorDiv((b - a) * t, PROVINCE_BLEND);
}

function buildLayers(target: Uint8Array, surface: Int16Array, mapSeed: number): void {
  for (let x = 0; x < W; x++) {
    const p = provinceAt(mapSeed, x);
    const pa = PROVINCES[p.a];
    const pb = PROVINCES[p.b];
    const bands: number[] = [];
    const leftRelief = iabs(surface[clampInt(x - 6, 0, W - 1)] - surface[x]);
    const rightRelief = iabs(surface[clampInt(x + 6, 0, W - 1)] - surface[x]);
    const relief = leftRelief > rightRelief ? leftRelief : rightRelief;
    const mantleScale = relief > 3 ? 0 : relief === 3 ? 16 : 64;
    for (let i = 0; i < 5; i++) bands.push(floorDiv(blendBand(pa.bands[i], pb.bands[i], p.t) * mantleScale, 64));
    const bedrockDepth = blendBand(pa.bedrockDepth, pb.bedrockDepth, p.t);

    let y = surface[x];
    const order = [SAND, SOIL, SAND, SCREE, SOIL];
    for (let i = 0; i < 5; i++) {
      if (bands[i] <= 0) continue;
      fillVertical(target, x, y, y + bands[i] - 1, order[i]);
      y += bands[i];
    }
    /* 기반암 상단. 선반 구역은 지표 가까이 올라온다 */
    let bedrockTop = BEDROCK_Y;
    if (bedrockDepth > 0) {
      const shelf = surface[x] + bedrockDepth;
      if (shelf < bedrockTop) bedrockTop = shelf;
    }
    if (y < bedrockTop) fillVertical(target, x, y, bedrockTop - 1, ROCK);
    fillVertical(target, x, bedrockTop, H - 1, BEDROCK);
  }
}

function addAnchors(target: Uint8Array, surface: Int16Array, mapSeed: number): void {
  for (let i = 0; i < ANCHOR_BASE_X.length; i++) {
    const jitter = (hash32(mapSeed, i, ANCHOR_SALT, 0) & 31) - 15;
    const x = clampInt(ANCHOR_BASE_X[i] + jitter, 24, 935);
    const top = clampInt(surface[x] + 96, 220, 500);
    for (let column = x - 3; column <= x + 2; column++) {
      fillVertical(target, column, top, H - 1, BEDROCK);
    }
  }
}

function addArch(target: Uint8Array, surface: Int16Array, mapSeed: number): void {
  const center = 480 + ((hash32(mapSeed, 0, ARCH_SALT, 0) & 127) - 63);
  const left = center - 72;
  const right = center + 72;
  let deepestSurface = surface[left];
  for (let x = left + 1; x <= right; x++) {
    if (surface[x] > deepestSurface) deepestSurface = surface[x];
  }
  const roofY = deepestSurface + 24;
  let legBottom = roofY + 64;
  const leftBottom = surface[left] + 112;
  const rightBottom = surface[right] + 112;
  if (leftBottom > legBottom) legBottom = leftBottom;
  if (rightBottom > legBottom) legBottom = rightBottom;
  legBottom = clampInt(legBottom, roofY + 64, 510);

  for (let x = left; x <= right; x++) fillVertical(target, x, roofY, roofY + 7, ROCK);
  for (let x = left + 8; x <= right - 8; x++) {
    fillVertical(target, x, roofY + 8, roofY + 56, EMPTY);
  }
  for (let x = left; x <= left + 7; x++) fillVertical(target, x, roofY, legBottom, ROCK);
  for (let x = right - 7; x <= right; x++) fillVertical(target, x, roofY, legBottom, ROCK);
}

export function surfaceCellY(source: Uint8Array, x: number): number {
  const column = clampInt(x, 0, W - 1);
  for (let y = 0; y < H; y++) {
    if (source[y * W + column] !== EMPTY) return y;
  }
  return H - 1;
}

function sealEdges(target: Uint8Array): void {
  const leftTop = surfaceCellY(target, 2);
  const rightTop = surfaceCellY(target, W - 3);
  for (let x = 0; x < 2; x++) fillVertical(target, x, leftTop, H - 1, BEDROCK);
  for (let x = W - 2; x < W; x++) fillVertical(target, x, rightTop, H - 1, BEDROCK);
}

export function buildMap(mapSeed: number): Uint8Array {
  const target = new Uint8Array(N);
  const surface = new Int16Array(W);
  for (let x = 0; x < W; x++) surface[x] = generatedSurface(mapSeed, x);
  buildLayers(target, surface, mapSeed);
  addAnchors(target, surface, mapSeed);
  addArch(target, surface, mapSeed);
  sealEdges(target);
  return target;
}

export function chooseSpawnCells(source: Uint8Array, playerCount: number, spawnSeed = 0): number[] {
  if (source.length !== N) throw new RangeError("grid length must be 960 * 540");
  if (playerCount < 2 || playerCount > 6) throw new RangeError("playerCount must be 2..6");

  const surface = new Int16Array(W);
  for (let column = 0; column < W; column++) surface[column] = surfaceCellY(source, column);
  const selected: number[] = [];
  const layout = hash32(spawnSeed, playerCount, 0x5a17, 0) % 3;
  const center = 80 + hash32(spawnSeed, 0, 0x5a17, 1) % (W - 160);
  const clusterRadius = 48 + playerCount * 24;
  for (let slot = 0; slot < playerCount; slot++) {
    const target = 32 + floorDiv(slot * (W - 64), playerCount - 1);
    let bestX = -1;
    let bestScore = 0x7fffffff;
    for (let searchPass = 0; searchPass < 2 && bestX < 0; searchPass++) {
      for (let column = 20; column < W - 20; column++) {
        if (surface[column] >= H - 12) continue;
        let separated = true;
        for (const other of selected) {
          if (iabs(column - other) < SPAWN_MIN_GAP) { separated = false; break; }
        }
        if (!separated) continue;
        const relief = iabs(surface[column - 6] - surface[column]) + iabs(surface[column + 6] - surface[column]);
        if (searchPass === 0 && relief > SPAWN_MAX_RELIEF) continue;
        const distance = layout === 0 ? iabs(column - center) - clusterRadius : layout === 1 ? iabs(column - target) - 64 : 0;
        const penalty = distance > 0 ? distance : 0;
        const score = penalty * 65536 + (hash32(spawnSeed, column, slot, 0x5a17) & 65535);
        if (score < bestScore) {
          bestScore = score;
          bestX = column;
        }
      }
    }
    if (bestX < 0) throw new RangeError("not enough separated ground for spawning");
    selected.push(bestX);
  }
  for (let index = selected.length - 1; index > 0; index--) {
    const other = hash32(spawnSeed, index, 0x5a17, 2) % (index + 1);
    const previous = selected[index]; selected[index] = selected[other]; selected[other] = previous;
  }
  return selected;
}

File: server/src/neodeol/sim/mapgen.py
"""결정론적 초기 맵 생성 — ``docs/mapgen.md`` 의 Python 미러."""

from __future__ import annotations

import numpy as np
import numpy.typing as npt

from neodeol import constants

from .intmath import clamp_int, floor_div, hash32_scalar, iabs
from .terrain import BEDROCK, EMPTY, H, N, ROCK, SAND, SCREE, SOIL, W

MAPGEN_VERSION = constants.MAPGEN_VERSION
NOISE_SHIFT = constants.NOISE_SHIFT
SURFACE_BASE = constants.SURFACE_BASE
SURFACE_AMP = constants.SURFACE_AMP
BEDROCK_Y = constants.MAPGEN_BEDROCK_Y
SPAWN_MIN_GAP = constants.SPAWN_MIN_GAP
SPAWN_MAX_RELIEF = constants.SPAWN_MAX_RELIEF

NOISE_MASK = 127
NOISE_SALT = 0x4D47
ANCHOR_SALT = 0xA11C
ARCH_SALT = 0xA2C4
ANCHOR_BASE_X = (160, 320, 640, 800)


def _fill_vertical(
    target: npt.NDArray[np.uint8],
    x: int,
    y_start: int,
    y_end: int,
    material: int,
) -> None:
    start = clamp_int(y_start, 0, H - 1)
    end = clamp_int(y_end, 0, H - 1)
    if start <= end:
        target[start : end + 1, x] = material


def _peak_at(column: int, center: int, width: int, amplitude: int) -> int:
    return clamp_int(width - iabs(column - center), 0, width) * amplitude // width


def _generated_surface(map_seed: int, column: int) -> int:
    sample = column >> NOISE_SHIFT
    fraction = column & NOISE_MASK
    start = hash32_scalar(map_seed, sample, NOISE_SALT, 0) & 0xFFFF
    end = hash32_scalar(map_seed, sample + 1, NOISE_SALT, 0) & 0xFFFF
    noise = start + (((end - start) * fraction) >> NOISE_SHIFT)
    main_seed = hash32_scalar(map_seed, 0, 0x504B, 0)
    side_seed = hash32_scalar(map_seed, 1, 0x504B, 0)
    valley_seed = hash32_scalar(map_seed, 2, 0x504B, 0)
    main_center = clamp_int(240 + (main_seed & 511), 240, 720)
    side_center = (770 if main_center < 480 else 190) + ((side_seed >> 16) & 63) - 31
    main_peak = _peak_at(column, main_center, 145 + ((main_seed >> 9) & 127), 148 + ((main_seed >> 16) & 63))
    side_peak = _peak_at(column, side_center, 120 + (side_seed & 63), 76 + ((side_seed >> 6) & 47))
    valley = _peak_at(column, (main_center + side_center) // 2, 110 + (valley_seed & 63), 32 + ((valley_seed >> 6) & 31))
    return clamp_int(SURFACE_BASE + (((noise - 32768) * SURFACE_AMP) >> 16) - main_peak - side_peak + valley, 72, 342)


# ══ 지질 구역 (mapgen.md §3.1) ═══════════════════════════════════════════
# `client/src/sim/mapgen.ts` 의 1:1 번역. 맵을 좌우로 나눠 구역마다 다른 지층을 깐다.
#
# 그 전에는 960열 전부가 같은 지층이라 안식각 26.6°/40.3°/44.2° 가 전술적으로 아무
# 일도 하지 않았다 — 어디를 파도 같은 것이 나오니 서 있는 자리가 의미가 없었다.

#: 구역 프로파일. `(bands, bedrock_depth)`.
#: bands 는 위에서부터 [SAND, SOIL, SAND, SCREE, SOIL] 두께이고 남는 깊이는 ROCK 이 채운다.
#: bedrock_depth 가 0 이면 `BEDROCK_Y` 를 그대로 쓴다.
#:
#: **표면 재질이 구역마다 다르다.** 앞 밴드를 0 으로 두어 아래 재질을 노출시킨다 —
#: 전부 SAND 로 시작하면 화면상 구분이 안 되고, 보이지 않으면 전술이 되지 않는다.
#: 지질 구역 프로파일. **값은 `constants.PROVINCE_TABLE` 이 유일한 사본이다** —
#: 여기 적으면 `SIM_VERSION` 이 안 따라온다. mapgen.md §4.1
PROVINCES: tuple[tuple[tuple[int, int, int, int, int], int], ...] = tuple(
    ((row[0], row[1], row[2], row[3], row[4]), row[5]) for row in constants.PROVINCE_TABLE
)

PROVINCE_SALT = 0x9E07
#: 구역 경계가 지층을 수직으로 자르지 않게 섞는 폭 (셀)
PROVINCE_BLEND = constants.PROVINCE_BLEND


def _province_order(map_seed: int) -> list[int]:
    """구역 배치 순열. **한 맵 안에 네 지질이 전부 나오는 것을 보장한다.**

    구역마다 독립 추첨하면 시드에 따라 맵 전체가 한 지질로 덮인다
    (실측: 모래 809열 / 960). 다양성이 이 변경의 전부라 우연에 맡기지 않는다.
    정수 전용 Fisher-Yates 이고 `hash32` 만 쓴다.
    """
    order = [0, 1, 2, 3]
    for i in range(len(order) - 1, 0, -1):
        j = hash32_scalar(map_seed, i, PROVINCE_SALT, 1) % (i + 1)
        order[i], order[j] = order[j], order[i]
    return order


def _province_at(map_seed: int, x: int) -> tuple[int, int, int]:
    """열 `x` 의 구역과 경계 혼합 가중치. 반환 `(a, b, t)`."""
    count = 4 + (hash32_scalar(map_seed, 0, PROVINCE_SALT, 0) & 1)
    width = W // count
    index = x // width
    if index >= count:
        index = count - 1
    local_x = x - index * width
    order = _province_order(map_seed)

    here = order[index % len(PROVINCES)]
    if local_x >= PROVINCE_BLEND or index == 0:
        return here, here, 0
    return order[(index - 1) % len(PROVINCES)], here, local_x


def _blend_band(a: int, b: int, t: int) -> int:
    """두 프로파일을 정수 가중 평균한다. `floor_div` 로 나눈다 (음수 대비)."""
    return a + floor_div((b - a) * t, PROVINCE_BLEND)


def _build_layers(
    target: npt.NDArray[np.uint8], surface: list[int], map_seed: int
) -> None:
    order = (SAND, SOIL, SAND, SCREE, SOIL)
    for x in range(W):
        pa_i, pb_i, t = _province_at(map_seed, x)
        pa_bands, pa_depth = PROVINCES[pa_i]
        pb_bands, pb_depth = PROVINCES[pb_i]
        relief = max(abs(surface[max(0, x - 6)] - surface[x]), abs(surface[min(W - 1, x + 6)] - surface[x]))
        mantle_scale = 0 if relief > 3 else 16 if relief == 3 else 64
        bands = [_blend_band(pa_bands[i], pb_bands[i], t) * mantle_scale // 64 for i in range(5)]
        bedrock_depth = _blend_band(pa_depth, pb_depth, t)

        y = surface[x]
        for i in range(5):
            if bands[i] <= 0:
                continue
            _fill_vertical(target, x, y, y + bands[i] - 1, order[i])
            y += bands[i]

        bedrock_top = BEDROCK_Y
        if bedrock_depth > 0:
            shelf = surface[x] + bedrock_depth
            if shelf < bedrock_top:
                bedrock_top = shelf
        if y < bedrock_top:
            _fill_vertical(target, x, y, bedrock_top - 1, ROCK)
        _fill_vertical(target, x, bedrock_top, H - 1, BEDROCK)


def _add_anchors(
    target: npt.NDArray[np.uint8], surface: list[int], map_seed: int
) -> None:
    for index, base_x in enumerate(ANCHOR_BASE_X):
        jitter = (hash32_scalar(map_seed, index, ANCHOR_SALT, 0) & 31) - 15
        x = clamp_int(base_x + jitter, 24, 935)
        top = clamp_int(surface[x] + 96, 220, 500)
        for column in range(x - 3, x + 3):
            _fill_vertical(target, column, top, H - 1, BEDROCK)


def _add_arch(target: npt.NDArray[np.uint8], surface: list[int], map_seed: int) -> None:
    center = 480 + ((hash32_scalar(map_seed, 0, ARCH_SALT, 0) & 127) - 63)
    left = center - 72
    right = center + 72
    deepest_surface = surface[left]
    for x in range(left + 1, right + 1):
        if surface[x] > deepest_surface:
            deepest_surface = surface[x]
    roof_y = deepest_surface + 24
    leg_bottom = roof_y + 64
    left_bottom = surface[left] + 112
    right_bottom = surface[right] + 112
    if left_bottom > leg_bottom:
        leg_bottom = left_bottom
    if right_bottom > leg_bottom:
        leg_bottom = right_bottom
    leg_bottom = clamp_int(leg_bottom, roof_y + 64, 510)

    for x in range(left, right + 1):
        _fill_vertical(target, x, roof_y, roof_y + 7, ROCK)
    for x in range(left + 8, right - 7):
        _fill_vertical(target, x, roof_y + 8, roof_y + 56, EMPTY)
    for x in range(left, left + 8):
        _fill_vertical(target, x, roof_y, leg_bottom, ROCK)
    for x in range(right - 7, right + 1):
        _fill_vertical(target, x, roof_y, leg_bottom, ROCK)


def surface_cell_y(source: npt.NDArray[np.uint8], x: int) -> int:
    column = clamp_int(x, 0, W - 1)
    source_2d = source.reshape(H, W)
    for y in range(H):
        if int(source_2d[y, column]) != EMPTY:
            return y
    return H - 1


def _seal_edges(target: npt.NDArray[np.uint8]) -> None:
    left_top = surface_cell_y(target, 2)
    right_top = surface_cell_y(target, W - 3)
    target_2d = target.reshape(H, W)
    for x in range(2):
        _fill_vertical(target_2d, x, left_top, H - 1, BEDROCK)
    for x in range(W - 2, W):
        _fill_vertical(target_2d, x, right_top, H - 1, BEDROCK)


def build_map(map_seed: int) -> npt.NDArray[np.uint8]:
    target = np.zeros((H, W), dtype=np.uint8)
    surface = [_generated_surface(map_seed, x) for x in range(W)]
    _build_layers(target, surface, map_seed)
    _add_anchors(target, surface, map_seed)
    _add_arch(target, surface, map_seed)
    _seal_edges(target.reshape(N))
    return target.reshape(N)


def choose_spawn_cells(source: npt.NDArray[np.uint8], player_count: int, spawn_seed: int = 0) -> list[int]:
    if source.size != N:
        raise ValueError("grid length must be 960 * 540")
    if player_count < 2 or player_count > 6:
        raise ValueError("player_count must be 2..6")

    surface = [surface_cell_y(source, column) for column in range(W)]
    selected: list[int] = []
    layout = hash32_scalar(spawn_seed, player_count, 0x5A17, 0) % 3
    center = 80 + hash32_scalar(spawn_seed, 0, 0x5A17, 1) % (W - 160)
    cluster_radius = 48 + player_count * 24
    for slot in range(player_count):
        target = 32 + slot * (W - 64) // (player_count - 1)
        best_x = -1
        best_score = 0x7FFFFFFF
        for search_pass in range(2):
            for column in range(20, W - 20):
                if surface[column] >= H - 12:
                    continue
                if any(iabs(column - other) < SPAWN_MIN_GAP for other in selected):
                    continue
                relief = iabs(surface[column - 6] - surface[column]) + iabs(surface[column + 6] - surface[column])
                if search_pass == 0 and relief > SPAWN_MAX_RELIEF:
                    continue
                distance = iabs(column - center) - cluster_radius if layout == 0 else iabs(column - target) - 64 if layout == 1 else 0
                score = max(0, distance) * 65536 + (hash32_scalar(spawn_seed, column, slot, 0x5A17) & 65535)
                if score < best_score:
                    best_score = score
                    best_x = column
            if best_x >= 0:
                break
        if best_x < 0:
            raise ValueError("not enough separated ground for spawning")
        selected.append(best_x)
    for index in range(len(selected) - 1, 0, -1):
        other = hash32_scalar(spawn_seed, index, 0x5A17, 2) % (index + 1)
        selected[index], selected[other] = selected[other], selected[index]
    return selected
